### PR TITLE
Fixes #24876 - Move repository sync to service class

### PR DESCRIPTION
**Pull request by sjha4**
To test run rake db:migrate and then test normal flows.

### CODE REVIEW

**Comment by jlsherrill** [comment]
This looks good!  i will give it a try

File: app/services/katello/pulp/repository.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+module Katello
+  module Pulp
+    class Repository < ::Actions::Pulp::Abstract
+      attr_accessor :repo, :input
+
+      def initialize(repo, input)
```
Comment (jlsherrill) [inline app/services/katello/pulp/repository.rb]: input here probably doesn't make much sense, instead you probably want to pass either a smart proxy, or a pulp_resource

File: app/services/katello/pulp/repository.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        @repo = repo
+        @input = input
+      end
+
+      def sync
```
Comment (jlsherrill) [inline app/services/katello/pulp/repository.rb]: I think this should pass in a hash of 'overrides', this would include a url override and 'validate'

File: app/services/katello/pulp/repository.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        sync_options = {}
+        sync_options[:max_speed] ||= SETTINGS[:katello][:pulp][:sync_KBlimit] if SETTINGS[:katello][:pulp][:sync_KBlimit]
+        sync_options[:num_threads] ||= SETTINGS[:katello][:pulp][:sync_threads] if SETTINGS[:katello][:pulp][:sync_threads]
+        sync_options[:feed] = @input[:source_url] if @input[:source_url]
+        sync_options[:validate] = !(SETTINGS[:katello][:pulp][:skip_checksum_validation])
```
Comment (jlsherrill) [inline app/services/katello/pulp/repository.rb]: I would use the dig method on these SETTINGS https://ruby-doc.org/core-2.3.0_preview1/Hash.html#method-i-dig

File: app/services/katello/pulp/repository.rb (hunk 2 above)
Comment (sjha4) [reply app/services/katello/pulp/repository.rb]: validate as in skip_checksum_validation setting?

File: app/services/katello/pulp/repository.rb (hunk 1 above)
Comment (sjha4) [reply app/services/katello/pulp/repository.rb]: We could remove the input here and just pass the Katello repository record?

File: app/services/katello/pulp/repository.rb (hunk 2 above)
Comment (jlsherrill) [reply app/services/katello/pulp/repository.rb]: yes, exactly, in fact all those 'sync_options' checks should be able to be passed in as a hash to override the default behavior.

**Comment by jlsherrill** [comment]
I would:
1) check out my newest commit in some branch
2) cherry pick your new commit ontop of it
3) git will have to correct the conflicts
4) fix the conflicts, commit, and force push over your existing branch

**Comment by sjha4** [comment]
[test katello]

File: app/lib/actions/pulp/abstract.rb [hunk 4]
```diff
@@ -5,7 +5,7 @@ class Abstract < Actions::Base
       middleware.use Actions::Middleware::PulpServicesCheck
 
       def pulp_resources(capsule_id = nil)
-        capsule_id ||= input["capsule_id"]
+        capsule_id ||= input["capsule_id"] if input
```
Comment (jlsherrill) [inline app/lib/actions/pulp/abstract.rb]: why is this required?  input should always exist

File: app/services/katello/pulp/repository.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        @repo = repo
+      end
+
+      def sync(overrides = {})
+        sync_options = {}
+        sync_options[:max_speed] = SETTINGS.dig(:katello, :pulp, :sync_KBlimit)
+        sync_options[:num_threads] = SETTINGS.dig(:katello, :pulp, :sync_threads)
+        sync_options[:feed] = overrides[:source_url] if overrides[:source_url]
+        sync_options[:validate] = !SETTINGS.dig(:katello, :pulp, :skip_checksum_validation)
+        sync_options.merge!(overrides[:options]) if overrides[:options]
```
Comment (jlsherrill) [inline app/services/katello/pulp/repository.rb]: I don't see how 'options' would ever get passed in here?

Comment (sjha4) [reply app/services/katello/pulp/repository.rb]: Updated to send input[:options] from the action..

File: app/lib/actions/pulp/repository/sync.rb [hunk 6]
```diff
@@ -16,25 +16,12 @@ def invoke_external_task
...
-            sync_options.merge!(input[:options]) if input[:options]
-
-            output[:pulp_tasks] = pulp_tasks =
-                                    [pulp_resources.repository.sync(input[:pulp_id], override_config: sync_options)]
-
+            overrides = {}
+            overrides[:feed] = input[:source_url] if input[:source_url]
+            overrides[:validate] = !(SETTINGS[:katello][:pulp][:skip_checksum_validation])
+            overrides[:options] = input[:options] if input[:options]
+            pulp_repo_params = {id: input[:pulp_id], resource: pulp_resources.repository}
```
Comment (sjha4) [inline app/lib/actions/pulp/repository/sync.rb]: Passing pulp resource and Id to the service class..Didn't think there's a way to get the pulp id from the resource..

**Comment by jlsherrill** [comment]
mind rebasing now that the main pr has been merged? :)

File: app/lib/actions/pulp/repository/sync.rb [hunk 7]
```diff
@@ -114,6 +101,11 @@ def ignored_tags
           # ignore background download tasks
           ["pulp:action:download"]
         end
+
+        def smart_proxy(capsule_id = nil)
+          smart_proxy = capsule_id ? SmartProxy.unscoped.find(input[:capsule_id]) : SmartProxy.first
+          smart_proxy
+        end
```
Comment (jlsherrill) [inline app/lib/actions/pulp/repository/sync.rb]: mind moving this to 'actions/pulp/abstract.rb' ?

I think you also need to check input[:capsule] here on line 106, not 'capsule_id'.   Also, i think you want to default to SmartProxy.default_capsule, not .first

File: app/services/katello/pulp/repository.rb [hunk 8]
```diff
@@ -0,0 +1,22 @@
+module Katello
+  module Pulp
+    class Repository < ::Actions::Pulp::Abstract
+      attr_accessor :pulp_repo_resource, :pulp_id
+
+      def initialize(pulp_repo_params)
+        @smart_proxy = pulp_repo_params[:smart_proxy]
+        @repo = pulp_repo_params[:repo]
```
Comment (jlsherrill) [inline app/services/katello/pulp/repository.rb]: this should accept a katello repository object, not some hash :)

File: app/services/katello/pulp/repository.rb [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+      end
+
+      def sync(overrides = {})
+        sync_options = {}
+        sync_options[:max_speed] = SETTINGS.dig(:katello, :pulp, :sync_KBlimit)
+        sync_options[:num_threads] = SETTINGS.dig(:katello, :pulp, :sync_threads)
+        sync_options[:feed] = overrides[:source_url] if overrides[:source_url]
+        sync_options[:validate] = !SETTINGS.dig(:katello, :pulp, :skip_checksum_validation)
+        sync_options.merge!(overrides[:options]) if overrides[:options]
+        [pulp_resources(overrides[:capsule_id]).repository.sync(@repo.pulp_id, override_config: sync_options.compact!)]
```
Comment (jlsherrill) [inline app/services/katello/pulp/repository.rb]: pulp_resources here won't work, will it?

Comment (sjha4) [reply app/services/katello/pulp/repository.rb]: We need a way to pass the capsule id to the abstract class..I saw something similar here : https://github.com/Katello/katello/blob/27d5de2493bfab921c3bf2060ea8f2fcfb364561/app/lib/actions/pulp/repository/refresh.rb#L11

Comment (jlsherrill) [reply app/services/katello/pulp/repository.rb]: that works because its in an action (and the pulp_extensions method is defined in the abstract pulp action).  Some of what we do is going to depend on what smart proxy we are used, so i think that should be part of the initializer:  

`def initialize(repo, smart_proxy)` and then we can reference it in the same way we reference the repo

**Comment by jlsherrill** [comment]
@sjha4 PR looks great!  Mind changing the commit message to leave out the 'pulp v3'.  Just 'moving repo sync to service class' should be fine.

**Comment by sjha4** [comment]
Merging..Thanks @jlsherrill!
